### PR TITLE
feat(fuzz): add XSS reflection context analyzer

### PR DESCRIPTION
**Pull request by ZachL111**
## feat(fuzz): add XSS reflection context analyzer

/claim #5838

### Proposed Changes

Adds an XSS reflection context analyzer to the fuzzing engine (`pkg/fuzz/analyzers/xss/`). Given an HTTP response body and a marker string, it classifies the HTML context of the reflection into one of 8 types (body text, attribute, URL attribute, event handler, executable script, non-executable script data, style, or comment). This gives the fuzzer the information it needs to select context-appropriate payloads instead of blindly spraying all of them. Relates to #5838.

### Problem

The fuzzing engine has no way to determine where in an HTML response a reflected value lands. Without this, XSS payload selection is blind, you either spray every payload everywhere (noisy, slow, false-positive-heavy) or miss valid injection points entirely. Issue #5838 asks for a context analyzer to fix this.

### What this PR does

Adds `pkg/fuzz/analyzers/xss/` — a standalone context classification package that takes a response body and a marker string, and returns which HTML context the marker was reflected into.

The function signature:

```go
func AnalyzeReflectionContext(responseBody, marker string) (XSSContext, error)
```

It uses `golang.org/x/net/html` tokenizer (already in `go.mod`) to walk the HTML token stream with a state machine tracking `inScript`, `inStyle`, and script executability. No regex for HTML parsing.

### Context types

| Context | When | Example |
|---------|------|---------|
| `ContextHTMLBody` | Text between tags | `<p>MARKER</p>` |
| `ContextHTMLAttribute` | Generic attribute | `<input value="MARKER">` |
| `ContextHTMLAttributeURL` | URL attribute (href, src, action, ping, etc.) | `<a href="/path/MARKER">` |
| `ContextHTMLAttributeEvent` | Event handler (onclick, onerror, etc.) | `<div onclick="fn(MARKER)">` |
| `ContextScript` | Executable `<script>`, or `javascript:`/`vbscript:`/`data:text/html`/`data:image/svg+xml` URI in an executable sink | `<a href="javascript:MARKER">` |
| `ContextScriptData` | Non-executable script (JSON, template, etc.) | `<script type="application/json">MARKER</script>` |
| `ContextStyle` | `<style>` block or `style=""` attribute | `<div style="color:MARKER">` |
| `ContextComment` | HTML comment | `<!-- MARKER -->` |

### Edge cases handled

These are the specific issues called out in community review of prior attempts:

- `javascript:` and `vbscript:` URIs classified as `ContextScript`, not `ContextHTMLAttributeURL`
- `data:text/html`, `data:application/xhtml+xml`, and `data:image/svg+xml` URIs classified as `ContextScript`
- Dangerous URI promotion is tag-specific — `<a href="javascript:...">` → `ContextScript`, but `` → `ContextHTMLAttributeURL` (browsers don't execute it)
- `<script type="application/json">` classified as `ContextScriptData`, not `ContextScript`
- Duplicate `type` attributes use first value per HTML5 spec (browsers ignore subsequent dupes)
- `srcdoc` attribute classified as `ContextHTMLBody` (renders full HTML)
- `style` attribute classified as `ContextStyle`, not `ContextHTMLAttribute`
- Event handlers (`onclick`, `onerror`, etc.) get their own `ContextHTMLAttributeEvent`
- Case-insensitive marker matching (servers may normalize case)
- No panics on malformed HTML, empty input, or binary data

### What this PR does NOT do

This is intentionally scoped to context classification only. It does not:
- Modify any existing files (`analyzers.go`, `request.go`, `http.go` — all untouched)
- Add new dependencies
- Send HTTP requests or inject canary payloads
- Implement the `Analyzer` interface (see "Integration path" below)

Prior PRs were likely not merged in part because they modified the shared `Options` struct in `analyzers.go` to add `ResponseBody`, renamed existing unexported functions, and mixed core context analysis with canary injection and payload replay logic. This PR avoids all of that.

### Integration path

The fuzzing pipeline already has the response body available as `bodyStr` at `pkg/protocols/http/request.go:985`, in scope when analyzers execute (line 1013). To wire this up:

1. Add a `ResponseBody string` field to `analyzers.Options`
2. Pass `bodyStr` into the options at the existing analyzer call site
3. The XSS analyzer's `Analyze()` method calls `AnalyzeReflectionContext(options.ResponseBody, marker)` and selects payloads based on context

### Usage example

```go
import "github.com/projectdiscovery/nuclei/v3/pkg/fuzz/analyzers/xss"

// responseBody is the HTML response from the server,
// marker is the unique string the fuzzer injected.
ctx, err := xss.AnalyzeReflectionContext(responseBody, marker)
if err != nil {
    log.Fatal(err)
}

switch ctx {
case xss.ContextScript:
    // use script breakout payloads
case xss.ContextHTMLAttribute:
    // use attribute escape payloads
case xss.ContextHTMLAttributeEvent:
    // use event handler payloads
case xss.ContextComment:
    // use comment breakout payloads
// ... etc
}
```

### Functional testing

To replicate and verify:

```sh
# clone and checkout the branch
git clone https://github.com/ZachL111/nuclei.git
cd nuclei
git checkout feat/xss-context-analyzer

# run the xss analyzer tests
go test ./pkg/fuzz/analyzers/xss/... -v -count=1

# verify no regressions in the rest of the fuzz package
go test ./pkg/fuzz/... -count=1

# build and vet
go build ./...
go vet ./pkg/fuzz/...
```

### Files added

```
pkg/fuzz/analyzers/xss/
├── context.go        — XSSContext type, iota constants, String()
├── analyzer.go       — AnalyzeReflectionContext() + helpers (~340 lines)
└── analyzer_test.go  — 63 table-driven test cases (~500 lines)
```

Zero files modified.

### Changelog

Changes made during review before submission:

1. **Fixed attribute consumption bug** — `TagAttr()` is a forward-only iterator on the tokenizer. The original code checked script type and marker in two separate loops, so the second loop saw no attributes. Merged both checks into a single `scanAttributes` pass. Added 3 test cases to cover `<script src="MARKER">` variations.
2. **Added `ping` to URL attributes** — missed on the first pass. The `ping` attribute on `<a>` tags fires a POST to the specified URL when clicked, so it's a valid URL injection context.
3. **Added `data:application/xhtml+xml` to dangerous URI detection** — `data:text/html` was covered but `data:application/xhtml+xml` renders and executes script the same way in iframes. Added a test case for it.
4. **Propagate real tokenizer errors** — `AnalyzeReflectionContext` was returning `nil` error on every `ErrorToken`, including actual parse failures. Now checks `tokenizer.Err()` and only swallows `io.EOF` (normal end of document). Real errors get surfaced to the caller.
5. **Removed dead code** — `foundType` bool in `scanAttributes` was redundant since `scriptType` already defaults to `""` which maps to executable in the lookup table. Cleaned it up.
6. **Added missing event handlers** — `onauxclick`, `onbeforeinput`, `onformdata`, `onslotchange`, `onsecuritypolicyviolation` were missing from the event handler set. Added them with test cases.
7. **Strip MIME parameters from script type** — `type="text/javascript; charset=utf-8"` was failing the exact lookup and getting misclassified as `ContextScriptData`. Now strips everything after `;` before checking. Added a test case.
8. **Fixed godoc comments on package-level vars** — `urlAttrs`, `eventHandlers`, `executableScriptTypes`, and `contextNames` had comments that didn't follow Go's godoc convention (comment must start with the entity name). Rewrote them so `go doc` and linters pick them up correctly.
9. **Added `data:image/svg+xml` to dangerous URI detection** — SVG data URIs can contain embedded JavaScript (`<svg onload=alert(1)>`) that executes when rendered in iframe/object/embed. Added test cases for both iframe (ContextScript) and img (ContextHTMLAttributeURL — browsers block SVG script execution in img tags).
10. **Fixed duplicate `type` attribute parser differential** — `scanAttributes` was overwriting `scriptType` on every `type` attribute encountered. HTML5 spec says browsers use the first attribute when dupes exist, so `<script type="application/json" type="text/javascript">` should be non-executable. Now only records the first `type`. Added a test case.
11. **Tag-specific dangerous URI classification** — `` was being classified as `ContextScript` even though browsers don't execute `javascript:` in img src. Added an `executableURLSinks` map that restricts `ContextScript` promotion to tag+attr pairs that actually execute (a+href, iframe+src, form+action, button+formaction, object+data, etc.). Everything else stays `ContextHTMLAttributeURL`. Added test cases for img src and ping with javascript: URIs.
12. **Added `vbscript:` URI detection** — covers IE11 and legacy Edge environments still deployed in corporate settings. Added a test case.
13. **Added `longdesc` to URL attributes** — `longdesc` on img/iframe elements can contain navigable URIs. Added a test case.

### Proof

```
$ go test ./pkg/fuzz/analyzers/xss/... -v -count=1

=== RUN   TestAnalyzeReflectionContext
=== RUN   TestAnalyzeReflectionContext/reflection_in_plain_HTML_body_text
=== RUN   TestAnalyzeReflectionContext/reflection_in_nested_div_body_text
=== RUN   TestAnalyzeReflectionContext/reflection_in_regular_attribute_value
=== RUN   TestAnalyzeReflectionContext/reflection_in_class_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_data-custom_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_title_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_href_with_regular_URL
=== RUN   TestAnalyzeReflectionContext/reflection_in_src_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_action_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_formaction_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_longdesc_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_onclick_handler
=== RUN   TestAnalyzeReflectionContext/reflection_in_onmouseover_handler
=== RUN   TestAnalyzeReflectionContext/reflection_in_onerror_handler
=== RUN   TestAnalyzeReflectionContext/reflection_in_onload_handler
=== RUN   TestAnalyzeReflectionContext/reflection_in_onauxclick_handler
=== RUN   TestAnalyzeReflectionContext/reflection_in_onbeforeinput_handler
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_block_with_no_type
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_type=text/javascript
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_type=module
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_type=application/javascript
=== RUN   TestAnalyzeReflectionContext/script_type_with_MIME_parameters_still_executable
=== RUN   TestAnalyzeReflectionContext/javascript_URI_in_href_must_be_ContextScript
=== RUN   TestAnalyzeReflectionContext/javascript_URI_with_whitespace_prefix
=== RUN   TestAnalyzeReflectionContext/javascript_URI_case-insensitive
=== RUN   TestAnalyzeReflectionContext/data:text/html_URI_in_src
=== RUN   TestAnalyzeReflectionContext/data:application/xhtml+xml_URI_in_src
=== RUN   TestAnalyzeReflectionContext/data:image/svg+xml_URI_in_iframe_src
=== RUN   TestAnalyzeReflectionContext/data:image/svg+xml_URI_in_img_src_does_not_execute
=== RUN   TestAnalyzeReflectionContext/vbscript_URI_in_href
=== RUN   TestAnalyzeReflectionContext/javascript_URI_in_img_src_does_not_execute
=== RUN   TestAnalyzeReflectionContext/javascript_URI_in_ping_does_not_execute
=== RUN   TestAnalyzeReflectionContext/reflection_in_ping_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_type=application/json
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_type=text/template
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_type=text/x-handlebars-template
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_type=application/ld+json
=== RUN   TestAnalyzeReflectionContext/duplicate_type_attributes_uses_first_per_HTML5_spec
=== RUN   TestAnalyzeReflectionContext/reflection_in_style_block
=== RUN   TestAnalyzeReflectionContext/reflection_in_style_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_HTML_comment
=== RUN   TestAnalyzeReflectionContext/reflection_in_comment_between_tags
=== RUN   TestAnalyzeReflectionContext/reflection_in_srcdoc_attribute
=== RUN   TestAnalyzeReflectionContext/case-insensitive_marker_matching_(lowercase_body)
=== RUN   TestAnalyzeReflectionContext/case-insensitive_marker_matching_(mixed_case_body)
=== RUN   TestAnalyzeReflectionContext/case-insensitive_in_attribute
=== RUN   TestAnalyzeReflectionContext/marker_not_found_in_response
=== RUN   TestAnalyzeReflectionContext/empty_response_body
=== RUN   TestAnalyzeReflectionContext/empty_marker
=== RUN   TestAnalyzeReflectionContext/malformed_HTML_with_unclosed_tags
=== RUN   TestAnalyzeReflectionContext/malformed_HTML_with_no_tags_at_all
=== RUN   TestAnalyzeReflectionContext/malformed_script_tag_not_closed
=== RUN   TestAnalyzeReflectionContext/broken_HTML_with_unclosed_attribute_quote
=== RUN   TestAnalyzeReflectionContext/broken_HTML_with_missing_closing_quote_but_valid_parse
=== RUN   TestAnalyzeReflectionContext/multiple_reflections_returns_first_context
=== RUN   TestAnalyzeReflectionContext/reflection_in_self-closing_tag_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_src_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_in_script_src_with_type_attribute
=== RUN   TestAnalyzeReflectionContext/script_tag_with_src_and_type_but_reflection_in_text
=== RUN   TestAnalyzeReflectionContext/non-executable_script_with_marker_in_src_attribute
=== RUN   TestAnalyzeReflectionContext/reflection_inside_noscript
=== RUN   TestAnalyzeReflectionContext/reflection_inside_textarea
--- PASS: TestAnalyzeReflectionContext (0.04s)
=== RUN   TestAnalyzeReflectionContext_NoPanic
--- PASS: TestAnalyzeReflectionContext_NoPanic (0.00s)
=== RUN   TestXSSContextString
--- PASS: TestXSSContextString (0.00s)
PASS
ok  	github.com/projectdiscovery/nuclei/v3/pkg/fuzz/analyzers/xss	0.510s

$ go build ./...    # zero errors
$ go vet ./pkg/fuzz/...    # zero warnings
```

### Checklist

- [x] Uses `golang.org/x/net/html` tokenizer (no regex)
- [x] No new dependencies
- [x] No existing files modified
- [x] All 8 context types detected correctly
- [x] All edge cases from prior PR reviews handled
- [x] Case-insensitive marker matching
- [x] No panics on malformed/empty/binary input
- [x] 63 table-driven tests, all passing
- [x] `go build ./...` passes
- [x] `go vet ./pkg/fuzz/...` passes
- [x] Existing tests unaffected
- [x] All godoc comments follow `// EntityName ...` convention

Closes #5838

### CODE REVIEW

**Comment by ZachL111** [comment]
I addressed all changes that were requested/commented from CodeRabbit, here's what I actually changed across both rounds of review:

1. **Fixed attribute consumption bug** — `TagAttr()` is forward-only on the tokenizer. The original code checked script type and marker in separate loops, so the second pass saw no attributes and `<script src="MARKER">` would silently miss the reflection. Merged both into a single `scanAttributes` pass. Added 3 test cases for `<script src="MARKER">` variations.

2. **Added `ping` to URL attributes** — missed it initially. `ping` on `<a>` tags fires a POST to the URL on click, so it's a valid URL injection context. Test case added.

3. **Added `data:application/xhtml+xml` to dangerous URI detection** — was only catching `data:text/html` but `xhtml+xml` renders and executes script the same way in iframes. Test case added.

4. **Error propagation** — `ErrorToken` now checks `tokenizer.Err()` and only swallows `io.EOF`. Actual parse errors get returned to the caller instead of silently returning `ContextUnknown, nil`. Kept the `error` in the signature since it's doing real work now.

5. **Dead code cleanup** — removed `foundType` and the redundant reset block in `scanAttributes`. `scriptType` already initializes to `""` which maps to executable in the lookup table, so the check was doing nothing.

53 tests still passing, build and vet clean.

Edit: Also I apologize if this is redundant, commenting a changelog here and in the original PR text, but I thought it best.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+		case html.ErrorToken:
+			// EOF is expected (end of doc), but surface real parse errors
+			if err := tokenizer.Err(); err != nil && err != io.EOF {
+				return ContextUnknown, err
+			}
+			return ContextUnknown, nil
+
+		case html.CommentToken:
+			if containsMarker(tokenizer.Token().Data, markerLower) {
+				return ContextComment, nil
```
[the remark this answers is not in the document]
Comment (ZachL111) [reply pkg/fuzz/analyzers/xss/analyzer.go:265]: Addressed all three:

**1. Missing event handlers** — Added `onauxclick`, `onbeforeinput`, `onformdata`, `onslotchange`, and `onsecuritypolicyviolation`. Added test cases for `onauxclick` and `onbeforeinput`.

**2. MIME parameters on script type** —  `type="text/javascript; charset=utf-8"` would fail the exact lookup and get classified as `ContextScriptData`. Now strips everything after `;` before the lookup. Added a test case for it.

**3. Return highest-priority context instead of first** — I hear the concern but I think this is a design decision rather than a bug. Returning the first match is intentional and predictable (there's a test case for it: "multiple reflections returns first context"). Defining a danger hierarchy adds opinions about what's "more exploitable" that belong in the caller, not the classifier. The cleaner path for this would be a follow-up that returns `[]XSSContext` with all reflection points and lets the fuzzer decide which to prioritize. Happy to do that as a separate PR if it'd be useful.

56 tests passing now versus 53 earlier.

[the remark this answers is not in the document]
Comment (ZachL111) [reply pkg/fuzz/analyzers/xss/analyzer.go:265]: Also fixed godoc comments on all package-level vars (`urlAttrs`, `eventHandlers`, `executableScriptTypes`, `contextNames`) as they weren't starting with the entity name, so linters flagged them. Purely cosmetic, no logic changes involved.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+		// also catch data:application/xhtml+xml as it renders and executes
+		// script the same way data:text/html does in iframes. Missed this one
+		// on the first pass.
+		if strings.HasPrefix(trimmed, "javascript:") ||
+			strings.HasPrefix(trimmed, "data:text/html") ||
+			strings.HasPrefix(trimmed, "data:application/xhtml+xml") {
+			return ContextScript
+		}
+		return ContextHTMLAttributeURL
+	}
```
[the remark this answers is not in the document]
Comment (ZachL111) [reply pkg/fuzz/analyzers/xss/analyzer.go:406]: I would agree, this was overclassifying. `` doesn't execute in any modern browser but was getting tagged as `ContextScript`, which would steer the fuzzer toward the wrong payloads.

Added an `executableURLSinks` map that restricts the `ContextScript` promotion to tag+attribute pairs that actually execute dangerous URIs in browsers:

- `href` on `a`, `area`
- `src` on `iframe`, `frame`, `embed`
- `action` on `form`
- `formaction` on `button`, `input`
- `data` on `object`

Everything else stays `ContextHTMLAttributeURL` even if the value is `javascript:` or `data:text/html`. Added test cases for `javascript:` in img src and in ping, with both correctly returning `ContextHTMLAttributeURL` now.

63 tests passing now.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+import (
+	"io"
+	"strings"
+
+	"golang.org/x/net/html"
+)
+
+// urlAttrs lists attributes whose values may contain navigable URIs.
+// ping was missed initially, it fires a POST to the URL when <a> is clicked.
+var urlAttrs = map[string]struct{}{
```
[the remark this answers is not in the document]
Comment (ZachL111) [reply pkg/fuzz/analyzers/xss/analyzer.go:81]: Added `longdesc` to the `urlAttrs` map. It's deprecated but still a valid URL attribute on img/iframe. Added a test case for it.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+	if attrName == "srcdoc" {
+		return ContextHTMLBody
+	}
+
+	if _, ok := urlAttrs[attrName]; ok {
+		trimmed := strings.TrimSpace(strings.ToLower(attrValue))
+		// also catch data:application/xhtml+xml as it renders and executes
+		// script the same way data:text/html does in iframes. Missed this one
+		// on the first pass.
+		if strings.HasPrefix(trimmed, "javascript:") ||
```
[the remark this answers is not in the document]
Comment (ZachL111) [reply pkg/fuzz/analyzers/xss/analyzer.go:391]: Good point to be fair. IE11 and legacy Edge are still around in corporate environments, many companies I contract for still use them. Added `vbscript:` to the dangerous URI prefix checks. Good to include. Added a test case for `<a href="vbscript:...">`.

[the remark this answers is not in the document]
Comment (ZachL111) [reply pkg/fuzz/analyzers/xss/analyzer.go:391]: Added `data:image/svg+xml` to the dangerous URI prefix checks. SVG data URIs can embed JavaScript via `<svg onload=...>` that executes when rendered in iframe/object/embed contexts.

Also made this tag-specific while I was at it, `<iframe src="data:image/svg+xml,...">` → `ContextScript`, but `` → `ContextHTMLAttributeURL` since browsers block SVG script execution in img tags. Added test cases for both.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+	var markerCtx XSSContext
+	markerFound := false
+	scriptType := ""
+
+	for {
+		key, val, more := tokenizer.TagAttr()
+		attrName := strings.ToLower(string(key))
+		attrValue := string(val)
+
+		if attrName == "type" {
```
[the remark this answers is not in the document]
Comment (ZachL111) [reply pkg/fuzz/analyzers/xss/analyzer.go]: `scanAttributes` was overwriting `scriptType` on every `type` attribute it saw, so the last one won. HTML5 spec says browsers use the first, `<script type="application/json" type="text/javascript">` is non-executable per the browser but the analyzer was calling it executable.

Fixed by adding a `foundType` bool that short-circuits after the first `type` attribute. Added a test case that verifies `<script type="application/json" type="text/javascript">` → `ContextScriptData`.

Should be the last of the issues, as far as I can see.

**Comment by ZachL111** [comment]
Hi @dogancanbakir and @Ice3man543, this PR implements the XSS context analyzer from the issue. All tests are good on my end, I kept implementing more so I can make sure this was all the way passing and complete. Would appreciate a review when you get a chance as I want to contribute to other issues as well, but I know in the guidelines it says one open PR at a time. Thanks!

**Review by Mzack9999 — APPROVED** [review]
